File: leetcode/java_unittest/distinctSubsequences_junit/distinctSubsequencesUnitTest.java
/*
 * int distinctSubsequences.numDistinct_01/02(String src, String pattern)
 * */

import org.junit.*;
import org.junit.Assert;

public class distinctSubsequencesUnitTest{
    private distinctSubsequences game;

    @Before
    public void setUp(){
        game = new distinctSubsequences();
    }

    @Test
    public void positiveTest(){
        String S = new String("rabbbit");
        String T = new String("rabit");
        int expected = 3;

        Assert.assertEquals(expected, game.numDistinct_01(S, T));
        Assert.assertEquals(expected, game.numDistinct_02(S, T));
    }

    @Test
    public void zeroTest(){
        String S = new String("ABB");
        String T = new String("ABBB");
        int expected = 0;

        Assert.assertEquals(expected, game.numDistinct_01(S, T));
        Assert.assertEquals(expected, game.numDistinct_02(S, T));
    }

    @Test
    public void negativeTest(){
        String S = new String("ABBCBBDBBF");
        String T = new String("ADC");
        int expected = 0;

        Assert.assertEquals(expected, game.numDistinct_01(S, T));
        Assert.assertEquals(expected, game.numDistinct_02(S, T));
    }

    @Test
    public void emptyTest(){
        String S = new String();
        String T = new String();
        int expected = 1;

        Assert.assertEquals(expected, game.numDistinct_01(S, T));
        Assert.assertEquals(expected, game.numDistinct_02(S, T));
    }

    @After
    public void settleDown(){
        game = null;
    }
}

File: leetcode/cpp_unittest/distinctSubsequences_unittest/distinctSubsequences_unittest.cpp
#include "distinctSubsequences.cpp"
#include "gtest/gtest.h"

/*
 * int Solution::numDistinct_01(const string& S, const string& T)
 *
 * int Solution::numDistinct_02(const string& S, const string& T)
 * */

TEST(numDistinct_01_Test, Positive01){
    Solution s;

    string S("rabbbit");
    string T("rabit");

    int expected = 3;
    EXPECT_EQ(expected, s.numDistinct_01(S, T));
}

TEST(numDistinct_01_Test, Negative01){
    Solution s;

    string S("ABCDE");
    string T("AEC");

    int expected = 0;
    EXPECT_EQ(expected, s.numDistinct_01(S, T));
}

TEST(numDistinct_02_Test, Positive01){
    Solution s;

    string S("rabbbit");
    string T("rabit");

    int expected = 3;
    EXPECT_EQ(expected, s.numDistinct_02(S, T));
}

TEST(numDistinct_02_Test, Negative01){
    Solution s;

    string S("ABCDE");
    string T("AEC");

    int expected = 0;
    EXPECT_EQ(expected, s.numDistinct_02(S, T));
}
